signac/core/file_buffered_synced_collection.py

```diff
@@ -0,0 +1,249 @@
+# Copyright (c) 2020 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+"""Implement the FileBufferedSyncedCollection class.
+
+The FileBufferedSyncedCollection is a concrete implementation of the buffer
+protocol established by the BufferedSyncedCollection class. It uses an
+in-memory cache to store data when in buffered mode. It is suitable for
+use with any file-based back end because it performs integrity checks based on
+whether or not the underlying file has been modified while buffering was
+activated.
+"""
+
+import errno
+import hashlib
+import json
+import os
+import sys
+
+from typing import Dict
+
+from .buffered_synced_collection import BufferedCollection
+from .caching import get_cache
+from .errors import MetadataError
+
+
+class FileBufferedCollection(BufferedCollection):
+    """Implement buffering for SyncedCollections with file-based backends.
+
+    All file-based backends can use the same set of integrity checks prior to a
+    buffer flush. This class standardizes that protocol.
+    """
+    _cache = get_cache()
+    _cached_collections: Dict[int, BufferedCollection] = {}
+    _BUFFER_CAPACITY = 32 * 2**20    # 32 MB
+    _CURRENT_BUFFER_SIZE = 0
+
+    def __init__(self, filename, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._filename = filename
+
+    @classmethod
+    def get_buffer_capacity(cls):
+        """Get the current buffer capacity."""
+        return cls._BUFFER_CAPACITY
+
+    @classmethod
+    def set_buffer_capacity(cls, new_capacity):
+        """Update the buffer capacity."""
+        cls._BUFFER_CAPACITY = new_capacity
+
+    @classmethod
+    def get_current_buffer_size(cls):
+        """Get the total amount of data currently stored in the buffer."""
+        return FileBufferedCollection._CURRENT_BUFFER_SIZE
```

Comment:
There's inconsistencies between `cls` and the explicit name `FileBufferedCollection`. Is this intentional (to control some behavior of subclasses) or should these all use `cls`? *This affects several places below, which I won't mark.*
```suggestion
        return cls._CURRENT_BUFFER_SIZE
```

Reply:
Yes there is a reason. At one point I was creating a separate cache for every subclass of this one, so that if (for example) JSONCollections and PickleCollections both created buffered versions, they would be stored in different classes. Once I implemented global buffering mode, though, I realized that this behavior could cause significant confusion with respect to things like setting the total buffer size. Would you be setting it per backend? What about between different data types (JSONList vs JSONDict)? I decided that the most natural expectation is that it would be truly global, so all backends would share one buffer, so that a user setting a particular buffer size wouldn't have to keep track of how many backends were available, or whether setting it for one backend would affect others. For that reason, I tried to replace all references to class variables relating to caching (the cache itself, its size and capacity, etc) with explicit references to FileBufferedSyncedCollection to avoid any ambiguity. 

It's entirely possible that I missed some references, so I'll do a quick review to make sure I changed it everywhere. I also should probably move the `[get|set]_buffer_capacity` methods out of the class entirely to avoid any implication that it's class-dependent (even if I make them staticmethods rather than classmethods it might be confusing).

Let me know if this rationale and the choices I made make sense to you. Happy to restructure if you think it can be improved.

Reply:
This design decision is fine with me, I just wanted to be sure that was intentional. It would help to have a comment in the code to indicate that the buffer is shared by all backends, since I wouldn't necessarily assume that behavior as a developer/user.

Reply:
Will do, I'll add that in the class docstring.